pkg/scheduler/queue/queue.go

```
@@ -245,41 +248,37 @@ func (q *RequestQueue) tryDispatchRequest(broker *queueBroker, call *nextRequest
 	}
 
 	call.lastUserIndex.last = idx
-	if queue == nil {
+	if req == nil {
 		// Nothing available for this querier, try again next time.
 		return false
 	}
 
-	// Pick next request from the queue. The queue is guaranteed not to be empty because we remove empty broker.
-	queueElement := queue.Front()
-
-	requestSent := call.send(nextRequestForQuerier{
-		req:           queueElement.Value,
+	reqForQuerier := nextRequestForQuerier{
+		req:           req,
 		lastUserIndex: call.lastUserIndex,
 		err:           nil,
-	})
+	}
+	requestSent := call.send(reqForQuerier)
 
 	if requestSent {
-		// If GetNextRequestForQuerier received the request, remove it from the queue.
-		// (GetNextRequestForQuerier might have already returned if its context was cancelled.)
-		queue.Remove(queueElement)
-
-		if queue.Len() == 0 {
-			broker.deleteQueue(tenantID)
-		}
-
 		q.queueLength.WithLabelValues(string(tenantID)).Dec()
+	} else {
+		if reqToEnqueue, ok := req.(requestToEnqueue); ok {
```

Would this conversion ever not succeed? I'm wondering if we should remove the guard and instead let it panic if this doesn't succeed, given it should never happen.

---

changed to panic with comment